### DISCUSSION TITLE
Custom keybindings support

### DISCUSSION THREAD

**Discussion by zachlloyd** [warpdotdev/Warp#500] [Archived - Launched]
Right now we only support hard-coded bindings.  We'd like to change this to support custom key maps.  Posting here as a tracking issue to get a sense of importance from the community.

Please upvote if you'd like this. See #579 for current status.

**Comment by elviskahoro** [comment]
Bundled keybinding profiles

Situation: I use tmux primarily for session management and would not mind using native session management as an alternative, if I'm able to keep the tmux bindings I'm used to. However, some bindings may clash with Warp's default keybindings, maybe in the config file there could be default keybinding profiles to help people ramp up and make a switch from their various and different tools.
Tmux is the first one that comes to mind for me, looking forward to ppl's feedback on this.

**Reply by Marvinus** [reply]
Would love to be able to use tmux key bindings in Warp to create/split panes and navigate.

**Reply by justdriveit** [reply]
Lol, Keyring is great.  I installed it, and my program changed it's file style on it's own, found random symbols to represent my launchers.  It was rather funny, but I think any Keyring is a pretty heavy risk.

**Comment by macgill4444** [comment]
+1 on `CMD M` for minimize window.

**Comment by tomtomdurrant** [comment]
+1 for custom keyboard shortcuts. 

In my current tmux setup I use `CTRL` + `HJKL` to move between splits, meaning that I don't have to move to the arrow keys. I don't find `CMD` + `OPT` + arrow very ergonomic.

**Comment by ZachFontenot** [comment]
I see there is already an issue tracking general Emacs bindings, but I just wanted to say it here. This is a pretty big dealbreaker for me, since I use it for almost every type of text interaction.

**Reply by beorn** [reply]
ctrl-p/n/f/b to navigate up/down/left/right (up/down in history and left/right on command line) are must-haves

**Comment by elviskahoro** [comment]
**CMD-D for split panes**

From @beorn via #600

Currently Warp uses cmd-e and cmd-shift-e for splitting panes, whereas the convention amongst terminals is to use cmd-d and cmd-shift-d.

It would be great if Warp was consistent with them.

Related to #542 which is asking to NOT use cmd-e for this.

**Reply by elviskahoro** [reply]
We're adding the ability to set custom keybindings, you'll be able to change it back. But the rationale for changing this is, we have VS Code keybindings for the input editor (fully fledged native text editor). And the input editor's keymaps take precedence in our view hierarchy i.e. CMD-D is for multi-cursor which doesn't apply in iTerm's case because the input does not support this kind of functionality.

It's worth changing your muscle memory in the long run for this. Here's a GIF from the docs. https://docs.warp.dev/features/input-text-editor

**Reply by elviskahoro** [reply]
From @beorn via #600 

On cmd-d: I understand, however I think you can make the case that for a terminal to be consistent with other terminals is probably more important than being consistent with a text editor. The case for multi-cursor is still somewhat unproven in a terminal (since you're 99% of the case just editing a single line), and there's always history substitution commands such as r old=new. Finally, split pane in vscode isn't cmd-e but cmd-\ right?

On ctrl-n/p/f/b: I guess these are still supported by even vscode out-of-the-box (and in every text field on

**Reply by elviskahoro** [reply]
so tldr; refactoring our keypress handler to support custom keybindings. once we finish we'll go through and add a bunch of missing keybindings.

we'll add the keybindings for partial completing when it's finished. we'll use fish keybindings, opt-f and ctrl-f. we don't want to add them now and then have to go through and re-add them using the new framework that supports devs setting custom mapping

**Comment by elviskahoro** [comment]
**CTRL-B - leader keymap similar to tmux**

From @kuator via: #6

If warp supports windows splitting and terminal multiplexing, it would be nice to have a <leader> keymapping a-la tmux. In tmux it is `ctrl-b` by default, but most people switch it to `ctrl-s` or `ctrl-a`.

**Reply by JesseFarebro** [reply]
Would love to see this, makes it really hard to switch over when I have muscle memory for lots of keybindings with a leader key. Some other terminals also support leader, e.g., WezTerm & Kitty.

**Comment by elviskahoro** [comment]
**ENTER - Allow configuring whether enter executes command or inserts new line (CMD-ENTER to run)**

From @colinking via #073

**Is your feature request related to a problem? Please describe.**
It would be great to be able to customize the `Enter` behavior in Warp.

By default, `Enter` executes a command while `Opt-Enter` inserts a newline. There's a similar UX in Slack, which they let you customize:

![image](https://user-images.githubusercontent.com/2907397/127178308-317cacd1-a505-42cd-9f7c-b900949b4466.png)

Would be great if a) this used `CMD` instead of `Opt` (for macOS at least) and b) if there was a way I could flip these bindings to match what I'm used to from Slack.

**Describe the solution you'd like**
(see above)

**Additional context**
Add any other context or screenshots about the feature request here.

**Reply by elviskahoro** [reply]
From @simjnd via #73
> A very common pattern on many apps/websites is to use `Shift` + `Enter` to insert a new line, so having the option to choose between `Alt`, `Cmd`, `Shift` and even `Ctrl` as a modifier alongside the `Enter` key could be a good move, offering more flexibility

**Reply by elviskahoro** [reply]
From @zheng via #73

> We want to support custom key-binding in general (see #30). This particular setting is really interesting! What would you say is the use case here? We take a lot of inspiration from Slack in general, and I can see this being useful for some segment of users.
> 
> By the way, both Shift + Enter and Opt + Enter will add a new line in Warp.

**Comment by elviskahoro** [comment]
**SHIFT-CMD-T - Reopen recently closed tab, pane, or window like in chrome**

From @gagata via #88:

**Describe the solution you'd like**
basic solution:
being able to reopen the closed tab at the previous pwd as it was before closing

killer feature:
restoring entire session from the given tab, including the history of executed commands, blocks etc.

**Additional context**
Add any other context or screenshots about the feature request here.

**Reply by elviskahoro** [reply]
From @Lucas-Kohorst via #88

also seems like iterm2 has some form of implementation of the killer feature https://iterm2.com/documentation-restoration.html

**Reply by elviskahoro** [reply]
From @gagata via #88

oh that's cool, thanks for the link! seems like they do this as an opt-in which makes a lot of sense

**Comment by elviskahoro** [comment]
**CTRL-N, CTRL-P to navigate lists**

From @blase via #89

**Is your feature request related to a problem? Please describe.**
In most software when presented with a list it's possible to move up and down using Ctrl-N (down) and Ctrl-(up). I think this is `readline` behavior and is not always consistent across operating systems, but most desktop apps in MacOS (especially the ones targeted at developers) follow that pattern as well.

**Describe the solution you'd like**
When presented with a list of choices I would like to navigate said list using Ctrl-N/Ctrl-P.

**Additional context**
<img width="366" alt="Screenshot 2021-08-03 at 09 55 03" src="https://user-images.githubusercontent.com/453270/127979973-fabcde18-6880-42f0-8dee-44a2763c7722.png">

**Reply by elviskahoro** [reply]
From @zheng via #89

> Makes sense, agree we should support this. It was also previously requested, just bumped the priority!

**Reply by elviskahoro** [reply]
From @ColmanHumphrey via #89 

> Similarly, I typically use Ctrl-p and -n to navigate history of commands instead of the arrow keys, so this would be great!

**Comment by elviskahoro** [comment]
**Support emacs keybindings throughout the terminal**

From @chriskrycho via #100

**Describe the bug**
Regardless of the setting of the "(Left|Right) Option Key is Meta", <kbd>⌥</kbd> rarely works for me.

**To Reproduce**
Steps to reproduce the behavior:
1. Enable either or both of the "(Left|Right) Option Key is Meta" options
2. Try to use <kbd>⌥</kbd>-powered commands for navigating, kill, yank, etc.
3. Observe that it doesn't consistently work.

Annoyingly, it works *sometimes*, but I haven't been able to identify under what conditions it does/n't work.

**Expected behavior**
Optimally, a Mac terminal editor should support *all* the features that CoreText supports out of the box. (I understand that that's a lot of work!)

**Screenshots**

Just to verify I have the setting turned on:

<img width="292" alt="Option Key is Meta" src="https://user-images.githubusercontent.com/2403023/128526715-983a504d-0dde-4bb0-9c3a-ef4d04cd1f00.png">

**Desktop (please complete the following information):**

- OS Version: macOS Version 11.5.1
- iMac (Retina 5K, 27-inch, 2019)

**Additional context**
Add any other context about the problem here.

**Reply by elviskahoro** [reply]
from @gagata via #100

Hi @chriskrycho! The recent stable release should contain the updated handling of the meta key. Since I was not able to repro your issue, do you mind giving it another try? I hope it's resolved, but let me know if not - happy to take another look at this

**Reply by elviskahoro** [reply]
from @chriskrycho via #100

Giving it a look now!

**Reply by elviskahoro** [reply]
from @chriskrycho via #100

Still broken, alas. Hitting <kbd>⌥</kbd><kbd>B</kbd>, which should be "move back by a word", just inserts ∫; and hitting <kbd>⌥</kbd><kbd>←</kbd>, which should *also* be "move back by a word", does nothing. Here's an example of the first:

https://user-images.githubusercontent.com/2403023/129575122-dff8e69c-6920-4876-ab0d-b249f388218f.mov

**Reply by elviskahoro** [reply]
from @gagata via #100 

ooh I see! thanks a lot for the video, it's super helpful - the fix enabled meta keybindings in emacs, but not in the rest of the terminal. will take a look at this this week then, just bear with me :)

**Reply by elviskahoro** [reply]
from @zachlloyd via #100

Changed the title to reflect what I think the underlying issue is, namely that we don't support emacs keybindings in our input area right now.

I think if you are in a focused emacs session, option as meta works as intended.

**Reply by elviskahoro** [reply]
from @zheng via #100 

We added a number of emacs bindings to all the editors in our app. We support most "motion" https://www.gnu.org/software/emacs/refcards/pdf/refcard.pdf keybindings now, but not the goto ones. From killing & deleting, we also do support C-y.

We specifically just added these motion keybindings ones in the last release:
meta-f
meta-backspace
meta-b
meta-a
meta-e
meta-shift-{
meta-shift-}
meta-shift-<
meta-shift->

**Reply by elviskahoro** [reply]
from @zheng via #100 

From rnp on discord:
It's very nice to see some emacs bindings, but they're really incomplete - I use ^B, ^F, ^P, ^N (and their various shift- and meta- variants), and none of those work.
^A, ^E, and ^K work as expected, so I find missing the back/fwd/prev/next bindings even more jarring

**Reply by elviskahoro** [reply]
from @matthewmturner via #100

@elviskahoro thanks for pointing me to this.  im a vim user and having similar issue.  in my case, im trying to use vim-tmux-navigator to seemless switch between vim and tmux pane (using Meta+h/j/k/l).  interestingly the meta mapping works when im in vim, but then once im in a tmux terminal pane it doesnt.

**Reply by WindSoilder** [reply]
Hi :-)

While `(Left|Right) Option Key is Meta` is on, I found that `Alt - b` works in local shell, but doesn't work inside `ipython/python`, or when you `ssh` to other machine, it doesn't work as well, the relative issue doesn't exist in iterm2.

I'm using `zsh`

**Comment by elviskahoro** [comment]
emacs

From @LitoMore via #292

### Describe the solution you'd like?

In most place of macOS, we can use Emacs-like keymaps to control our cursor. Such as moving cursor, deleting characters...

But in Warp. I cannot use <kbd>control + n</kbd> and <kbd>control + p</kbd> to replace arrow keys.

Here is some information about the macOS built-in Emacs-like keymaps: https://jblevins.org/log/kbd

Here is a simple list to let you understand the habits of Emacs users:

> _`Any textarea` means any textarea on macOS, such as the textarea in browser, Notes app..._

| Keymap | Function | Any textarea | iTerm / Terminal.app | Warp |
|:-:|:-:|:-:|:-:|:-:|
| <kbd>control + n</kbd> | Down arrow | Yes | Yes | No |
| <kbd>control + p</kbd> | Up arrow | Yes | Yes | No |
| <kbd>control + b</kbd> | Left arrow | Yes | Yes | No |
| <kbd>control + f</kbd> | Right arrow | Yes | Yes | No |
| <kbd>control + t</kbd> | Switch characters | Yes | Yes | No |
| <kbd>control + d</kbd> | Delete right side character | Yes | Yes | No |
| <kbd>control + g</kbd> | Cancel / ESC | Yes | Yes | No |
| <kbd>control + h</kbd> | Delete left side character | Yes | Yes | Yes |
| <kbd>control + a</kbd> | Move to start of line | Yes | Yes | Yes |
| <kbd>control + e</kbd> | Move to end of line | Yes | Yes | Yes |
| <kbd>control + u</kbd> | Delete line | Yes | Yes | Yes |
| <kbd>control + j</kbd> | Return | No | Yes | No |

### Is your feature request related to a problem? Please describe.

I strongly recommend finding an Emacs and Vim expert in your team to be a consultant for this feature.

If there is not. I would like to help.


### Additional context

_No response_

**Reply by elviskahoro** [reply]
From @srl295 via - #292

complete blocker for me. I had to repeat my workflow several times in slow motion to figure out what of my ingrained muscle memory was telling Warp to suddenly shut the tab.

Control-P, Control-A, Control-F, Control-D

I'll come back and try Warp again when this is fixed!

**Reply by srl295** [reply]
Yea and for the record it's fixed enough that I'm using warp again.

**Comment by elviskahoro** [comment]
**CMD-left/right to switch terminal tabs**

From @benwaffle via #162

### Describe the solution you'd like?

```shell
Pressing ⌘+⬅️ should go to the previous tab, and ⌘+➡️ should go to the next tab. iTerm 2 has this.
```

### Is your feature request related to a problem? Please describe.

_No response_

### Additional context

_No response_

**Reply by elviskahoro** [reply]
From @benwaffle via #162 

^+➡️ and ^+⬆+➡️ doesn't seem to be working for me

**Reply by elviskahoro** [reply]
From @zheng via #162 

For this, we have:
* ^ + tab and ^ + shift + tab (native mac shortcut)
* cmd + # like in chrome tabs

Are these all iTerm2 specific shortcuts?

**Reply by elviskahoro** [reply]
From @benwaffle via #162 

Looks like iTerm 2 supports all 3 ways

**Reply by elviskahoro** [reply]
From @JanJakes via #162

> cmd + # like in chrome tabs

@zheng This doesn't seem to work for keyboards that use numeric keys for special symbols and numbers themselves are done using `Shift + #` (like the Czech keyboard). In iTerm, Chrome, and other apps `CMD + #` works for me even on the Czech keyboard. Should I create a new issue for that?

**Reply by elviskahoro** [reply]
From @ephemer via #162

Please leave `Cmd + Left` / `Cmd + Right` as they are :( In a text editor (which a terminal largely is, IMHO) these shortcuts move to the start / end of a line. I would really like it to stay that way.

`Ctrl (+ Shift) + Tab` is the standard for switching tabs and I also think it should stay that way 👍🏼

**Comment by elviskahoro** [comment]
**CMD-# -> Ability to switch panes instead of tabs**

From @crayment via #240 

### Describe the solution you'd like?

```shell
Ability to bind Command + 1 to select the first pane, etc.
```


### Is your feature request related to a problem? Please describe.

```shell
iTerm has a setting for "Shortcut to choose a split pane" which I have set to command + number and it's deeply engraved in my muscle memory at this point.
```


### Additional context

<img width="796" alt="Screen Shot 2021-10-01 at 11 29 10 AM" src="https://user-images.githubusercontent.com/78082/135662572-c8e97fda-f99b-4657-ae48-2f1492dce5e0.png">

**Reply by elviskahoro** [reply]
From @kevinyang372

Thanks for reporting! Definitely agree we need to build better navigation for panes. Will reach back once this is built!

**Comment by elviskahoro** [comment]
**CTRL-SHIFT-UP/DOWN to add cursors**

From @ephemer via #374

### Describe the solution you'd like?


In Sublime Text (and VSCode I think) you can add an additional cursor in the line above/below the current cursor(s) via `Ctrl + Shift + Up` / `Ctrl + Shift + Down`.

To me that is a fundamental part of multi-cursor support, equally important to or even more important than `Cmd + D`. It would be fantastic if you could add this.



### Is your feature request related to a problem? Please describe.


Selecting a word and duplicating that selection is more keystrokes and effort than just adding a cursor directly. I don't always want to add cursors on lines that contain the same word / command.

From what I can see Ctrl+Shift+Down (etc.) are not bound to anything else and implementing this would be a huge fluidity win for transferring my mental model of multi-cursor editing into the terminal.

**Comment by beorn** [comment]
Control a-z:

- ctrl-a - move beg of line
- ctrl-b - move back one char (MISSING)
- ctrl-c - interrupt
- ctrl-d - delete forward one char (MISSING) - if on empty command line = exit term
- ctrl-e - move to end of line
- ctrl-f - move forward on char (MISSING)
- ctrl-g - visual bell (MISSING?)
- ctrl-h - delete back one char = backspace
- ctrl-i - 
- ctrl-j - line feed (MISSING) - useful for editing multiline commands without executing them
- ctrl-k - kill to end of line (kill buffer)
- ctrl-l - clear screen
- ctrl-m - carriage return (MISSING - not needed?)
- ctrl-n - move down / next in history (MISSING)
- ctrl-o - accept current line and queue next line in history (MISSING) - run after history search or history browse
- ctrl-p - move up / prev in history (MISSING)
- ctrl-q - continue after term output stop (after ctrl-s)
- ctrl-r - search history backwards
- ctrl-s - stop term output (continue with ctrl-q) - or history search forward
- ctrl-t - transpose chars (MISSING)
- ctrl-u - kill backward beg of line
- ctrl-v - makes next character types verbatim (quotes it) (MISSING?)
- ctrl-w - kill word backward (kill buffer)
- ctrl-x - 
- ctrl-y - yank kill buffer
- ctrl-z - suspend process

So missing ctrl-b/d/f/g/j/n/p/t/v

**Reply by elviskahoro** [reply]
thanks @beorn I'm gonna add them to our list:
- https://github.com/warpdotdev/Warp/issues/579

**Reply by elviskahoro** [reply]
finished updating it!

**Reply by a-takahashi223** [reply]
> ctrl-u - kill backward beg of line

Version v0.2022.04.18.09.08.stable_04, ctrl-u "copy and clear selected line." How can I reset to "kill backward beg of line?"

**Reply by a-takahashi223** [reply]
In https://github.com/warpdotdev/Warp/discussions/500#discussioncomment-1980049 , ctrl-u "Delete line." Is this one adopted?
If so, how can I disable "copy and clear selected line" (like "Check for updates" which is not assigned any keyboard shortcuts by default)?

**Comment by elviskahoro** [comment]
**ALT-. -> iterate over the arguments of prior commands**

From @meredian via #461

### Describe the solution you'd like?

I liked experience from iTerm:
Use (on mac) `Option + .` to loop through last argument from historical commands & copy them to command line. On each invocation it goes deeper into history & overwrites previously injected values.

Very handy for running various commands on same file or quickly iterate through historically used files (most often they're passed as last argument)

#### Examples of behaviour:

```
git diff some-file
git diff some-other-file
pwd
git add [Option + .] pwd
        [Option + .] some-other-file
        [Option + .] some-file
```

### Is your feature request related to a problem? Please describe.

Running various commands over same file is an often pattern - saves you from memorising exact name, especially if you touch multiple files concurrently.

**Comment by elviskahoro** [comment]
**CTRL-LEFT/RIGHT to partially complete autosuggestion**

From @itsrifat via #488

### Describe the solution you'd like?

zsh autosuggestion allows you to complete the auto suggestion partially with `ctrl + right-arrow` one word at a time. It would be nice to have the feature in warp. Right now you have to accept the whole command with the press of a a right arrow

![zsh-autosuggestion](https://user-images.githubusercontent.com/4328772/144669778-d87ad6aa-5faa-4081-aead-01464b77dc27.gif)



### Is your feature request related to a problem? Please describe.

_No response_

### Additional context

_No response_

**Comment by elviskahoro** [comment]
**OPT -> in alt-screens to bypass mouse selection (vim)**

From @meredian via #584

### Describe the solution you'd like?

```shell
In order to make sure that you can copy on-screen text in any mode, add button which while being pressed allows to select any text with mouse on terminal screen as-is (and copy it to buffer).

E.g on Mac in iTerm you can select any on screen by pressing [Option] key - that works for Vim & all other cases as well
```


### Is your feature request related to a problem? Please describe.

```shell
Normally you can select any part of on-screen text and copy it to system buffer.

When having Vim with `set mouse=a` it becomes impossible to select text in terminal screen since vim intercepts selection & turns on visual mode. So no way you can just select-and-copy select without diving into vim yanking.
```


### Additional context

_No response_

**Comment by elviskahoro** [comment]
**CTRL-G to exit command history search**

From @d-lord via #594

### Describe the solution you'd like?

I'd love to copy the Emacs convention and allow Ctrl+G to cancel the Ctrl+R history search. You already support Esc and Ctrl+C, of course.


### Is your feature request related to a problem? Please describe.

Just muscle memory for people used to emacs/fzf/many other implementers of readline keybindings.


### Additional context

[fzf already binds this by default](https://github.com/junegunn/fzf/blob/master/README.md?plain=1#L208).

[A notable readline cheatsheet](https://readline.kablamo.org/emacs.html) mentions it too.

This might end up under #579.

PS - just started trying out Warp today. So far, incredible! Thanks!

**Reply by elviskahoro** [reply]
From @arora-aditya via #594

I think a more over-arching solution is to allow for the command palette commands to be configurable by the user (eg: a different mapping for copying the command output like ctrl + cmd + C instead of cmd + shift + option + c)

**Comment by elviskahoro** [comment]
> Is there a way to enable esc+. on Warp, like to insert the last arg of the previous command? I usually bind it to alt+. in my terminals. Very handy for when you want to do two actions on the same thing (like create and cd to a dir, or mv something then cd to the dir you moved it to)

**Reply by elviskahoro** [reply]
https://unix.stackexchange.com/questions/260125/esc-what-is-it-and-where-i-can-know-more-about-it

**Comment by elviskahoro** [comment]
From @saket via #298

Hey folks, I'm not sure why but I'm unable to navigate between words using option-arrow keys. I checked that it's a [supported shortcut](https://docs.warp.dev/features/keyboard-shortcuts#input-editor-shortcuts) so I'm not sure why is it not working for me. How can I help you debug this?

**Reply by elviskahoro** [reply]
From aloke
Hey there! Would you mind checking if you have "option-as-meta" enabled? 
![image](https://user-images.githubusercontent.com/4110292/138613520-cb0dbee3-e532-4725-ae42-abb5606bc5ef.png). That could be one reason why this isn't working

**Reply by elviskahoro** [reply]
From @saket

Hey Alok, they're both disabled on my side.

<img width="282" alt="Screen Shot 2021-10-24 at 8 44 54 PM" src="https://user-images.githubusercontent.com/2387680/138619627-a19521ab-665b-428a-b803-881187594179.png">

**Reply by elviskahoro** [reply]
From @none-da
I am on 2015 Macbook Pro, running latest warp (v0.2021.11.09.19.46.stable_01) and "Left option key is Meta" is enabled, still "option+left/right arrows" are not working as expected.

**Comment by rrrrando** [comment]
iTerm2 has this neat feature to show and hide all windows with one hotkey or key combination:

<img width="785" alt="image" src="https://user-images.githubusercontent.com/34184401/151370163-a1f71247-53e6-4cbf-9629-48d1b52f09d3.png">

I use the `§` key and kept typing that char everywhere at first.

**Reply by elviskahoro** [reply]
@rrrrando See:
- #91

**Comment by tombaker** [comment]
In iTerm2, after `set -o vi`, I can do `Esc` `kkk` to call previous commands to the command line, then `$`, `^`, `w`, etc to navigate within the command, `i` to enter insert mode, etc.

Warp designers may have a better way to do this, but I have been using vim (vi) since the early 1980s, and I can play it using muscle memory the way one would play an oboe. At my age, I would rather play that oboe well than be handed, say, a saxophone. If the Warp orchestra could accommodate an oboe, I could continue to make good music while extending my range with split panes and the rest. I'm just one data point, but lack of a vim mode would be a dealbreaker for me.

**Reply by ttnppedr** [reply]
@tombaker Here is a new issue for vim mode #867

**Comment by elviskahoro** [comment]
Hey everyone!

We recently launched an update with support for setting custom keybindings. We have [directions on our documentation.](https://docs.warp.dev/features/keyboard-shortcuts)

Will share bugs that have been reported and updates as we iterate.

For the emacs keymap check out this repo:
https://github.com/warpdotdev/keysets

**Reply by LitoMore** [reply]
But it still does not support Emacs keymap.

**Reply by tombaker** [reply]
Or Vim

**Reply by elviskahoro** [reply]
@LitoMore For emacs keymap check this out:
https://github.com/warpdotdev/keysets/tree/main

@tombaker We do not have modal support yet within Warp for keybindings. Will share updates when this is worked on.

**Comment by WindSoilder** [comment]
While `(Left|Right) Option Key is Meta is on`, I found that `Alt - b` works in local shell, but doesn't work inside `ipython/python`

and `Alt - backspace` doesn't works in `ipython/python` and [helix editor](https://github.com/helix-editor/helix)(What's strange is that it insert `ackspace` into the editor) the relative issue doesn't exist in iterm2.

I'm using zsh

**Comment by carlosefonseca** [comment]
I'm new on Warp.

I can't seem to do, on the prompt, ESC+. to fill the last argument of the previous command.

I also can't seem to manage to get Alt+Left/Right to work on prompts like `pry` or `ipython` (It adds characters instead of moving the cursor). I believe that on iTerm I have set custom hex codes for those keys. Is it/will it be possible to have this somehow?
![CleanShot 2022-04-06 at 10 16 33](https://user-images.githubusercontent.com/14120/161941515-3530c346-19b9-4cbb-b87b-997438325e14.png)
I can do ESC+B/F but I have to hit ESC again to repeat which is annoying. I did try to change the keybinding for moving one word but didn't seem to work.

**Comment by warpdotdev-devx** [comment]
Can you try navigating to Settings > Features > and toggle the "Left Option key is Meta"?

**Comment by bilsch** [comment]
Not sure if this is the right place to ask so let me know if there is another/better issue/way!

I use vi mode on the command-line. Things like escape **h|j|k|l,** vi style editing on the cli either inline or in an $EDITOR etc. I don't seem to have this option and did not find anything directly. Is there such a thing? I think the idea is related - ditto with tmux type sequences.

**Reply by bsergean** [reply]
I have the same problem, `set -o vi` is what I have in my .zshrc and I've been using this for years, so going back to the emacs terminal editing is awkward.

**Comment by aggKartik** [comment]
I'm not able to option + arrows keys to go between words with custom key bindings on latest warp on rails console still. Was anyone able to figure out a solution that I could use?

**Comment by tdterry** [comment]
I don't know if this is the correct place to post this or not. I found that ALT+LEFT/RIGHT/BACKSPACE do not work in subshells. ALT+LEFT/RIGHT does nothing. ALT+BACKSPACE is a normal backspace. I found this when running the mongo shell, but I also reproduced it with a plain bash subshell.

**Comment by dannyneira** [comment]
Discord user shmike#0653 requested:
> Allow the Keybind of `CMD-K` which is set to "close current session" to be changed to something else.

**Comment by NxSoftware** [comment]
Please allow overriding the `CMD+T` key bind for opening a new tab. I prefer working with panes and was previously using this shortcut to open a new pane in iTerm2, along with `CMD+SHIFT+T` to split the pane in the other orientation (basically the same as `CMD+D` and `CMD+SHIFT+D` does in Warp)

**Comment by ErikSom** [comment]
I'm so used to pressing TAB to autocomplete stuff, it would be nice if I could swap the current right arrow key with the tab.

**Comment by wsrast** [comment]
I find that the Warp terminal eats many of my Neovim key commands with no way to have those keystrokes pass through. I use `ctrl+h,j,k,l` as my window navigation directions in Neovim, and simply cannot do that in Warp. I have to spawn Neovide as a separate app in order to get full control of my editor away from Warp. This seems like a major oversight.

I thought I read somewhere that there's a configuration file where you can declare "passthroughKeys" in JSON, but for the life of me, I can no longer find these anywhere.

**Reply by zachbai** [reply]
Hey @wsrast,  Warp developer + fellow neovim user here. I use the same keybindings to navigate between buffers in nvim and don't encounter any issue within Warp (without config of `passthroughKeys`).

It's possible there are some Warp keyboard shortcuts bound to those keystrokes that are "eating" them and prevent them from being sent to nvim -- to check/play around with this, go to Settings > Keyboard shortcuts; you can search for existing keyboard shortcuts using the name of the keystrokes in the search bar e.g. "control h" to search for shortcuts bound to ctrl+h

**Comment by gplusplus314** [comment]
Suggestion: **Allow `Command` to be mapped to `Control`**

I'm giving Warp a shot, coming from Alacritty. The most immediate workflow showstopper I'm experiencing is the lack of ability to effectively use the Command key as the Control key within the terminal. This is a muscle memory and ergonomic optimization that I can't live without and is one of the primary reasons why I use Alacritty.

**"But why?" - you, maybe. Stay with me.**

There are four modifiers on a keyboard. Desktop operating systems agree on their use cases, but Apple's macOS swaps two modifiers by convention. At an abstract level, all the major OS keyboard idioms have a modifier for in-focus hotkeys (do something in the window that is currently in focus) and global hotkeys. These are commonly the Control and GUI/Super modifiers on most operating systems, including the BSDs, which explains why terminal applications (shells, Vim, etc) expect this modifier for hotkeys. But this creates a problem in Mac: the idiom is backwards.

Generally speaking, macOS reverses the role of the Control and GUI/Super (Command in macOS parlance) modifiers, making Control the least used modifier, lending itself as a popular choice for global keyboard shortcuts, such as switching virtual desktops. In fact, Apple's keyboard layout moves the in-focus hotkey modifier to the left thumb for even better ergonomics.

But this idiom is broken in most terminals on macOS. This is especially difficult for people who switch between macOS and other operating systems as well as [people who rely on ergonomic keyboards](https://github.com/gerryhernandez/gkey).

This effect is achieved in Alacritty by [binding every combination](https://github.com/gerryhernandez/gdot/blob/main/.config/alacritty/macos.yml#L85) of `Command+<Key>` to sending a control-character to the terminal. While it's not the most elegant solution, it works. Warp could do something like this, or even better, offer a boolean toggle to do it automatically, similar to what it already does with the `Option->Meta` remap.

My suggestion lands somewhere in the intersection of accessibility, customization, and keybinds, so I figured this was a decent thread to chime in on. Thanks for reading my giant wall of text!

**Reply by andreiborisov** [reply]
Since Warp focuses on productivity and UX so much, this should be somewhat of a priority. It's much more idiomatic to use something like `CMD-Escape` instead of `Ctrl-C` for interrupting a running command.

**Reply by mieubrisse** [reply]
FWIW you can use Karabiner to do exactly this for you. I've found it much easier than trying to get all the individual programs you want to build the necessary rebinding features. E.g. I've rebound Ctrl-J to ↑ and Ctrl-K to ↓ across all apps

**Comment by karambaq** [comment]
Still no solution for Option+key/backspace in subshells like python?

**Reply by miranquil** [reply]
Sigh, it seems so.
At lease I'm sure that `Option+Backspace` in Helix editor inserts `ackspace` for now. (when "Left Option key is Meta ON)

**Reply by WindSoilder** [reply]
My biggest issue for using warp instead of iterm2/wezterm is the issue like this

**Reply by miranquil** [reply]
Same as me...

**Comment by x-tropy** [comment]
Could Warp show which key binding is conflicting, just like what vscode did. In such way, I can quickly locate the conflicting items.

**Comment by mark-jordanovic-lewis** [comment]
Coming from tmux I would love to have the option for leader keys. Making the switch to warp on MacOs was smooth. On my i3 machine my fingers get confused and I am often closing warp rather than splitting the pane.

**Comment by yoshintame** [comment]
Is there solution for Option/Shift/Cmd+arrow/delete in subshells?

**Comment by s2xon** [comment]
Switching to ghosttly till this is added 🙏

**Reply by cage-is** [reply]
[Ghostty*](https://ghostty.org)
Just adding for the uninitiated and avoid confusion.
